core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java

```diff
@@ -1281,13 +1349,314 @@ public String toStringResponseOnly(Function<? super ResponseHeaders, ?> headersS
         return sanitized != null ? sanitized.toString() : "<sanitized>";
     }
 
-    private static final class ListenerEntry {
-        final RequestLogListener listener;
+    private static final class RequestLogFuture extends CompletableFuture<RequestLog> {
         final int interestedFlags;
 
-        ListenerEntry(RequestLogListener listener, int interestedFlags) {
-            this.listener = listener;
+        RequestLogFuture(int interestedFlags) {
             this.interestedFlags = interestedFlags;
         }
+
+        void completeLog(RequestLog log) {
+            super.complete(log);
+        }
+
+        @Override
+        public boolean complete(RequestLog value) {
+            // Disallow users from completing arbitrarily.
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public boolean completeExceptionally(Throwable ex) {
+            // Disallow users from completing arbitrarily.
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public boolean cancel(boolean mayInterruptIfRunning) {
+            return false;
+        }
+
+        @Override
+        public void obtrudeValue(RequestLog value) {
+            // Disallow users from completing arbitrarily.
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void obtrudeException(Throwable ex) {
+            // Disallow users from completing arbitrarily.
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    /**
+     * A {@link RequestLog} that delegates to {@link DefaultRequestLog} to read properties without
+     * checking availability.
+     */
+    private final class CompleteRequestLog implements RequestLog {
+        @Override
+        public boolean isComplete() {
+            return DefaultRequestLog.this.isComplete();
+        }
+
+        @Override
+        public boolean isRequestComplete() {
+            return DefaultRequestLog.this.isRequestComplete();
+        }
+
+        @Override
+        public boolean isAvailable(RequestLogProperty property) {
+            return DefaultRequestLog.this.isAvailable(property);
```

Comment:
```suggestion
            return true;
```

Reply:
Just realized this is not possible because we return `notCheckingAccess` when `isRequestComplete() == true` as well. I think we need to choose between:

1. provide two different complete log implementations (`CompleteRequestLog` and `CompleteRequestOnlyLog`). Lazy-allocate `CompleteRequestOnlyLog`.
2. do not pass `notCheckingAccess` to `requestCompleteFuture` and optimize for `completeFuture`.
3. pass `notCheckingAccess` to both `requestCompleteFuture` and `completeFuture`.

We are currently at (3), but I find it somewhat confusing to work with. How about (1) or (2)? (1) sounds good to me. What do you think?